sw/airborne/modules/ctrl/object_tracking.c

```diff
@@ -0,0 +1,161 @@
+ * @author Gautier Hattenberger <gautier.hattenberger@enac.fr>
+ * Control a rotorcraft heading to track an object detected by a camera
+ */
+
+#include "modules/ctrl/object_tracking.h"
+
+#include "firmwares/rotorcraft/navigation.h"
+#include "subsystems/abi.h"
+#include "generated/airframe.h"
+#include "generated/modules.h"
+
+// ABI message binding ID
+#ifndef OBJECT_TRACKING_ID
+#define OBJECT_TRACKING_ID ABI_BROADCAST
+#endif
+
+// Timout in seconds before entering search mode
+#ifndef OBJECT_TRACKING_TIMEOUT
+#define OBJECT_TRACKING_TIMEOUT 3.0f
+#endif
+
+// Deadband (constant heading setpoint)
+#ifndef OBJECT_TRACKING_DEADBAND
+#define OBJECT_TRACKING_DEADBAND RadOfDeg(5)
+#endif
+
+// Turn rate in tracking mode (rad/s)
+#ifndef OBJECT_TRACKING_RATE
+#define OBJECT_TRACKING_RATE RadOfDeg(10)
+#endif
+
+// Turn rate in search mode (rad/s)
+#ifndef OBJECT_TRACKING_SEARCH_RATE
+#define OBJECT_TRACKING_SEARCH_RATE RadOfDeg(20)
+#endif
+
+// Select rate or heading tracking mode
+#ifndef OBJECT_TRACKING_USE_RATE
+#define OBJECT_TRACKING_USE_RATE FALSE
+#endif
+
+// Send debug message
+#ifndef OBJECT_TRACKING_DEBUG
+#define OBJECT_TRACKING_DEBUG TRUE
+#endif
+
+#if OBJECT_TRACKING_DEBUG
+#include "subsystems/datalink/downlink.h"
+#include "pprzlink/messages.h"
+#include "mcu_periph/uart.h"
+#endif
+
+int16_t object_tracking_deadband;
+float object_tracking_rate;
+float object_tracking_search_rate;
+
+float object_lat;
+float object_vert;
+float timeout;
+
+abi_event object_ev;
+
+static const float nav_dt = 1. / NAV_FREQ;
+
+// callback on follow me message
+static void get_object(uint8_t sender_id __attribute__((unused)),
+    uint32_t id __attribute__((unused)),
+    uint8_t frame __attribute__((unused)),
+    float heading, float height,
+    float distance __attribute__((unused)))
+{
+  object_lat = heading;
+  object_vert = height;
+  timeout = 0.f;
+}
+
+void object_tracking_init(void)
+{
+  object_tracking_deadband = OBJECT_TRACKING_DEADBAND;
+  object_tracking_search_rate = OBJECT_TRACKING_SEARCH_RATE;
+  object_tracking_rate = OBJECT_TRACKING_RATE;
+
+  object_lat = 0.f;
+  object_vert = 0.f;
+  timeout = OBJECT_TRACKING_TIMEOUT; // start in search mode
+
+  // Bind to camera message
+  AbiBindMsgFOLLOW_ME(OBJECT_TRACKING_ID, &object_ev, get_object);
+}
+
+void object_tracking_run(void)
+{
+#if OBJECT_TRACKING_USE_RATE
+  float rate = object_tracking_search_rate; // search mode
+  if (timeout < OBJECT_TRACKING_TIMEOUT) {
+    timeout += nav_dt;
+    if (object_lat > object_tracking_deadband) {
+      rate = object_tracking_rate;
+    } else if (object_lat < -object_tracking_deadband) {
+      rate = -object_tracking_rate;
+    } else {
+      rate = 0.f;
+    }
+#if OBJECT_TRACKING_DEBUG
+    float msg[] = {
+      rate,
+      object_lat,
+      object_vert,
+      timeout
+    };
+    DOWNLINK_SEND_PAYLOAD_FLOAT(DefaultChannel, DefaultDevice, 4, msg);
+#endif
+  }
+  nav_heading += ANGLE_BFP_OF_REAL(rate * nav_dt);
```

Comment:
I still think yo should bound this in this function. Add INT32_COURSE_NORMALIZE(nav_heading) at the end of the function.

Another thing is I am not sure it is a great idea to increment the nav_heading set-point using a fixed rate when getting the target from a camera pointing in the real vehicle heading. Imagine the case when the vehicle can't track the heading set-point defined by nav_heading, then nav_heading will increment faster than the vehicle and probably lead to overshooting the target. It would seem safer to use the true heading as a reference, something like: 

`nav_set_heading_deg(stateGetNedToBodyEulers_f()->psi + rate * nav_dt)`

Alternatively, you can limit the maximum angle diff between nav_heading and the current heading to make sure it doesn't run away, similar to stabilization with STABILIZATION_ATTITUDE_SP_PSI_DELTA_LIMIT. You could even use the object_tracking_deadband or (object_tracking_deadband / 2) to limit it, then the overshoot should be limited to within the deadband.

Reply:
I removed the rate mode for now, it wasn't performing well (a real PID might help, but I don't have the time to do it now)